### PR TITLE
time: Make TimeZoneRule blittable and avoid copies

### PR DESCRIPTION
**Pull request by Thog**
This drastically reduce overhead of using TimeZoneRule around the codebase by avoiding to copy TimeZoneRule 16384 bytes all around.

Effect on games is unknown but could improve perf a bit depending on JIT output.

### CODE REVIEW

File: Ryujinx.HLE/HOS/Services/Time/StaticService/ITimeZoneServiceForPsc.cs [hunk 1]
```diff
@@ -200,9 +203,9 @@ public ResultCode ToCalendarTime(ServiceCtx context)
                 throw new InvalidOperationException();
             }
 
-            TimeZoneRule rules = MemoryHelper.Read<TimeZoneRule>(context.Memory, bufferPosition);
+            ref TimeZoneRule rules = ref context.Memory.GetRef<TimeZoneRule>(bufferPosition);
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/StaticService/ITimeZoneServiceForPsc.cs:206]: This should use `GetSpan` as `GetRef` will throw if the memory region is not contiguous.

File: Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneContentManager.cs [hunk 2]
```diff
@@ -146,7 +146,11 @@ public IEnumerable<(int Offset, string Location, string Abbr)> ParseTzOffsets()
                         continue;
                     }
 
-                    TimeZone.ParseTimeZoneBinary(out TimeZoneRule tzRule, tzif.Get.AsStream());
+                    Common.Memory.Box<TimeZoneRule> tzRuleBox = new Common.Memory.Box<TimeZoneRule>();
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneContentManager.cs:149]: I'd add `using Ryujinx.Common.Memory` to avoid fully specifying the namespace here.

File: Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneManager.cs [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
 using System.IO;
-using static Ryujinx.HLE.HOS.Services.Time.TimeZone.TimeZoneRule;
 
 namespace Ryujinx.HLE.HOS.Services.Time.TimeZone
 {
     class TimeZoneManager
     {
         private bool                 _isInitialized;
-        private TimeZoneRule         _myRules;
+        private Box<TimeZoneRule>    _myRules;
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneManager.cs:11]: Why is that necessary? `_myRules` is already on the heap even without the `Box` since `TimeZoneManager` is a reference type.

File: Ryujinx.HLE/HOS/Services/Time/StaticService/ITimeZoneServiceForPsc.cs (hunk 1 above)
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Time/StaticService/ITimeZoneServiceForPsc.cs:206]: Just to clarify, `GetSpan` does not throw because it returns a `ReadOnlySpan`. It assumes you don't need to modify the data, so it can just return a copy if the memory region is not contiguous. `GetRef` on the other range, does assume that you might modify the data, and so if the memory region is not contiguous, it can't return the ref as it wouldn't point to the whole data.

File: Ryujinx.Common/Memory/Box.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿namespace Ryujinx.Common.Memory
+{
+    public class Box<T> where T : unmanaged
+    {
+        private T _data;
+
+        public ref T Data => ref _data;
```
[the remark this answers is not in the document]
Comment (Thog) [reply Ryujinx.Common/Memory/Box.cs:7]: I just want to avoid the field being accessed in a non ref way and initially I was using an array.

I could try changing that if you want

File: Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneContentManager.cs (hunk 2 above)
Comment (Thog) [reply Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneContentManager.cs:149]: It was conflicting with LibHac Box, will force a using Box =

File: Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneManager.cs (hunk 3 above)
Comment (Thog) [reply Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneManager.cs:11]: Easier to replace later on when loading a timezone to replace it without making a full copy

File: Ryujinx.Common/Memory/Box.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gdkchan) [reply Ryujinx.Common/Memory/Box.cs:7]: You can still access 'Data' without 'ref' and it will do a copy of the value in this case.
Example:
```cs
Box<MyStruct> box = new Box<MyStruct>();
DoSomethingWithMyStruct(box.Data); // This does a copy.
DoSomethingWithMyStructByRef(ref box.Data); // This doesn't.
```
Both returning a ref of the field and making the field public and accessing it directly would give the same result here.

File: Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneManager.cs (hunk 3 above)
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneManager.cs:11]: You can just replace all instances of `ref _myRules.Data` with `ref _myRules`, and on `SetDeviceLocationNameWithTimeZoneRule` you can do `bool timeZoneConversionSuccess = TimeZone.ParseTimeZoneBinary(ref _myRules, timeZoneBinaryStream);` and remove `Box<TimeZoneRule> rules = new Box<TimeZoneRule>();` and `_myRules            = rules;`. It will achieve the same result and get rid of the `Box<T>`.

Comment (Thog) [reply Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneManager.cs:11]: I cannot use _myRules for ParseTimeZoneBinary as this doesn't account for possible parsing issues, that also doesn't account for temporary TimeZoneRule used for posix name parsing

Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZoneManager.cs:11]: Right, I see what you mean now, you don't want `_myRules` to be left in a invalid state in the case the function fails.

File: Ryujinx.Common/Memory/Box.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gdkchan) [reply Ryujinx.Common/Memory/Box.cs:7]: I think it's fine to leave it like this so marking as resolved.

File: Ryujinx.HLE/HOS/Services/Time/TimeZone/Types/TimeZoneRule.cs [hunk 5]
```diff
@@ -22,17 +24,32 @@ struct TimeZoneRule
...
+        private AtsStorageStruct _ats;
 
-        [MarshalAs(UnmanagedType.ByValArray, SizeConst = TzMaxTypes)]
-        public TimeTypeInfo[] Ttis;
+        public Span<long> Ats => SpanHelpers.AsSpan<AtsStorageStruct, long>(ref _ats);
 
-        [MarshalAs(UnmanagedType.ByValArray, SizeConst = TzCharsArraySize)]
-        public char[] Chars;
+        [StructLayout(LayoutKind.Sequential, Size = sizeof(byte) * TzMaxTimes)]
+        private struct TypesStorageStruct {}
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/TimeZone/Types/TimeZoneRule.cs:35]:
```suggestion
        private struct TypesStorageStruct { }
```

File: Ryujinx.HLE/HOS/Services/Time/TimeZone/Types/TimeZoneRule.cs [hunk 6]
```diff
@@ -22,17 +24,32 @@ struct TimeZoneRule
...
+
+        [StructLayout(LayoutKind.Sequential, Size = TimeTypeInfo.Size * TzMaxTimes)]
+        private struct TimeTypeInfoStorageStruct { }
+
+        private TimeTypeInfoStorageStruct _ttis;
+
+        public Span<TimeTypeInfo> Ttis => SpanHelpers.AsSpan<TimeTypeInfoStorageStruct, TimeTypeInfo>(ref _ttis);
+
+        [StructLayout(LayoutKind.Sequential, Size = sizeof(byte) * TzCharsArraySize)]
+        private struct CharsStorageStruct {}
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/TimeZone/Types/TimeZoneRule.cs:49]:
```suggestion
        private struct CharsStorageStruct { }
```

**Review by gdkchan — COMMENTED** [review]
nit, just for consistency.

File: Ryujinx.HLE/HOS/Services/Time/TimeZone/Types/TimeZoneRule.cs [hunk 7]
```diff
@@ -22,17 +24,32 @@ struct TimeZoneRule
         [MarshalAs(UnmanagedType.I1)]
         public bool GoAhead;
 
-        [MarshalAs(UnmanagedType.ByValArray, SizeConst = TzMaxTimes)]
-        public long[] Ats;
+        [StructLayout(LayoutKind.Sequential, Size = sizeof(long) * TzMaxTimes)]
+        private struct AtsStorageStruct {}
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Time/TimeZone/Types/TimeZoneRule.cs:28]:
```suggestion
        private struct AtsStorageStruct { }
```
Sorry, I missed this one on last review.

**Review by gdkchan — APPROVED** [review]
Other than the comment, looks good. Would be nice if we had some test coverage for this code.

**Review by riperiperi — APPROVED** [review]
LGTM

**Comment by gdkchan** [comment]
Needs a rebase now.

File: Ryujinx.HLE/HOS/Services/Time/TimeZone/Types/TimeZoneRule.cs (hunk 7 above)
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Time/TimeZone/Types/TimeZoneRule.cs:28]: This and the other comments were not addressed (or maybe they were and somehow got reverted with the rebase?)

Comment (Thog) [reply Ryujinx.HLE/HOS/Services/Time/TimeZone/Types/TimeZoneRule.cs:28]: huh odd? I was pretty sure I did address that one

Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Time/TimeZone/Types/TimeZoneRule.cs:28]: If you applied the suggestion from GitHub, then I guess you forgot to pull the changes on your local branch and they got lost on rebase.
